File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        if(head==null || head.next==null){
            return false;
        }
        ListNode prev = head;
        ListNode curr = head.next;
        while(curr!=null && curr.next!=null){
            if(prev.equals(curr)){
                return true;
            }
            prev = prev.next;
            curr = curr.next.next;
        }
        return false;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if head.next is None:
            return 
        if head.next.next is None:
            return
        if head.next.next.next is None:
            second = head.next
            last = head.next.next
            head.next = last
            last.next = second
            second.next = None
            return
        first = head
        last = first
        second_last = None
        while last.next is not None:
            second_last = last
            last = last.next
        second = first.next
        first.next = last
        second_last.next = None
        self.reorderList(second)
        last.next = second
        